### DISCUSSION TITLE
RuCore / Расширения / Core_Ex_Address_Fias

### DISCUSSION THREAD

**Discussion by eugenekogan-spb** [fhir-ru/core#238] [Минимальный набор артефактов]
//--------------------------------------
//страница: Руководство RuCore / Расширения / Core_Ex_Address_Fias
//--------------------------------------

```
Alias: $Ядро_Расширение_КодыФИАС = Core_Ex_Address_Fias
Extension: Core_Ex_Address_Fias
Id:        core-ex-address-fias
Title: "Кодирование адреса по ФИАС"
Description: "Кодирование адреса по ФИАС. При маппинге на СЭМД данный url соответствует urn:hl7-ru:fias"

* extension contains
    aoguid 1..1 MS and
    houseguid 0..1 MS

* extension[aoguid]
  * value[x] only Identifier
  * valueIdentifier
    * system = "urn:hl7-ru:fias:aoguid"
    * value 1..1 MS
    * value ^short = "GUID адресного объекта."

* extension[houseguid]
  * value[x] only Identifier
  * valueIdentifier
    * system = "urn:hl7-ru:fias:houseguid"
    * value 1..1 MS
    * value ^short = "GUID дома."
```
Пример заполнения адреса: https://github.com/fhir-ru/core/discussions/237#discussion-5772797
Страница с методическими рекомендациями по описанию Extension: https://github.com/fhir-ru/core/discussions/246

**Comment by Vrodvi** [comment]
1. В идентификаторе не обозначено требование на указание пространства идентификаторов ФИАС 

Следует указать то пространство, которое указано в специально для этого созданном ресурсе - https://github.com/fhir-ru/core/discussions/242.

**Reply by Vrodvi** [reply]
Требование на указание пространства идентификаторов добавлено.
Замечание "1." закрыто.

**Comment by Vrodvi** [comment]
2. Кардинальность значения позволяет не указывать идентификатор

Кардинальность позволяет не указать идентификатор, но при этом нет никакого аналога nullFlavor или ещё каких-либо способов обозначить причину отсутствия идентификатора.

Более того в СЭМДах идентификатор адресного объекта вообще не может быть не указан - у него требование ОБЯЗАН и даже nullFlavor использовать не разрешено. А для ситуации когда идентифицировать адрес по ФИАСу в целом не возможно предусмотрено указание nullFlavor для всего комплексного расширения.

Какую замену сообщество предлагает вместо этих возможностей?

**Reply by Vrodvi** [reply]
Замечание "2." не закрыто.
Всё ещё нет решения о необходимости указания специального значения при отсутствии данных (в данном случае - для идентификатора дома):
- требуется ли всегда указывать все элементы структуры (с добавлением специального атрибута на подобие nullFlavor)? Или считаем, что будет достаточно пропуска опциональных атрибутов?

---

Вариант (1):
- Атрибутов nullFlavor не нужно, достаточно опциональности атрибутов.
  - При отсутствии идентификатора дома по ФИАС - пропускается атрибут houseguid комплексного расширения.
  - При отсутствии любых идентификаторов ФИАС - пропускается всё комплексное расширение

--- 

Вариант (2):
- Атрибутов nullFlavor нужен, но только для вложенных атрибутов внутри комплексного расширения ФИАС.
  - При отсутствии идентификатора дома по ФИАС - атрибут houseguid комплексного расширения остаётся, но в нём указывается nullFlavor.
  - При отсутствии любых идентификаторов ФИАС - пропускается всё комплексное расширение
```
Extension: Core_Ex_Address_Fias
Id: core-ex-address-fias
Title: "Кодирование адреса по ФИАС"
Description: "Кодирование адреса по ФИАС."
* ^url = "urn:hl7-ru:fias"

* obeys fias-house-or-nullflavor

* extension contains
    aoguid 1..1 MS and
    houseguid 0..1 MS

* extension[aoguid].value[x] only Identifier
* extension[aoguid].valueIdentifier.system = "urn:hl7-ru:fias:aoguid"
* extension[aoguid].valueIdentifier.value 1..1 MS
* extension[aoguid].valueIdentifier.value ^short = "GUID адресного объекта"

* extension[houseguid].value[x] only Identifier
* extension[houseguid].valueIdentifier.system = "urn:hl7-ru:fias:houseguid"
* extension[houseguid].valueIdentifier.value 0..1 MS
* extension[houseguid].valueIdentifier.value ^short = "GUID дома"
* extension[houseguid].valueIdentifier.extension contains http://hl7.org/fhir/StructureDefinition/iso21090-nullFlavor named nullFlavor 0..1 MS

Invariant:   fias-house-or-nullflavor
Description: "Должны быть указаны либо данные по GUID дома из ФИАС либо nullFlavor."
Expression: "(extension[houseguid].valueIdentifier.value.exists() or extension[houseguid].valueIdentifier.extension[nullFlavor].exists()) and (extension[houseguid].valueIdentifier.value.exists() and extension[houseguid].valueIdentifier.extension[nullFlavor].exists()).not()"
Severity: #error
```
(плюс нужна проверка, что не указано и значение GUID дома и nullFlavor одновременно - это делается через Invariant - я привёл пример такого ограничения, но нужно проверить насколько правильно оно описано и работает ли это наложенное ограничение)

---

Вариант (3):
- Атрибут nullFlavor нужен, и для вложенных атрибутов внутри комплексного расширения ФИАС и для самого комплексного расширения ФИАС.
  - При отсутствии идентификатора дома по ФИАС - атрибут houseguid комплексного расширения остаётся, но в нём указывается nullFlavor.
  - При отсутствии любых идентификаторов ФИАС - комплексное расширение остаётся, но в но в нём указывает nullFlavor.
```
Extension: Core_Ex_Address_Fias
Id: core-ex-address-fias
Title: "Кодирование адреса по ФИАС"
Description: "Кодирование адреса по ФИАС."
* ^url = "urn:hl7-ru:fias"

* obeys fias-or-nullflavor
* obeys fias-house-or-nullflavor

* extension contains
    aoguid 1..1 MS and
    houseguid 0..1 MS and
    http://hl7.org/fhir/StructureDefinition/iso21090-nullFlavor named nullFlavor 0..1 MS

* extension[aoguid].value[x] only Identifier
* extension[aoguid].valueIdentifier.system = "urn:hl7-ru:fias:aoguid"
* extension[aoguid].valueIdentifier.value 1..1 MS
* extension[aoguid].valueIdentifier.value ^short = "GUID адресного объекта"

* extension[houseguid].value[x] only Identifier
* extension[houseguid].valueIdentifier.system = "urn:hl7-ru:fias:houseguid"
* extension[houseguid].valueIdentifier.value 0..1 MS
* extension[houseguid].valueIdentifier.value ^short = "GUID дома"
* extension[houseguid].valueIdentifier.extension contains http://hl7.org/fhir/StructureDefinition/iso21090-nullFlavor named nullFlavor 0..1 MS

Invariant:   fias-or-nullflavor
Description: "Должны быть указаны либо данные по ФИАС либо nullFlavor."
Expression: "(extension[aoguid].exists() or extension[nullFlavor].exists()) and (extension[aoguid].exists() and extension[nullFlavor].exists()).not()"
Severity: #error

Invariant:   fias-house-or-nullflavor
Description: "Должны быть указаны либо данные по GUID дома из ФИАС либо nullFlavor."
Expression: "(extension[houseguid].valueIdentifier.value.exists() or extension[houseguid].valueIdentifier.extension[nullFlavor].exists()) and (extension[houseguid].valueIdentifier.value.exists() and extension[houseguid].valueIdentifier.extension[nullFlavor].exists()).not()"
Severity: #error
```
(плюс нужна проверка, что не указано и значение GUID дома и nullFlavor одновременно или значения GUIDов и nullFlavor самого комплексного расширения - это делается через Invariant - я привёл пример такого ограничения, но нужно проверить насколько правильно оно описано и работает ли это наложенное ограничение)
Плюс в этом варианте необходимо установить данное комплексное расширение как обязательное для указания (в типе данных Адрес):
```
Profile: CoreAddress
Id: core-address
Parent: Address
Title: "Адрес (Ru Core)"
Description: "Адрес"

// ... не относящееся к вопросу наполнение вырезано для краткости

* extension contains 
  $Ядро_Расширение_КодыФИАС named fias 1..1 MS

* extension[fias] ^short = "Коды по ФИАС"
```

**Reply by Vrodvi** [reply]
Замечание не закрыто - требуется принять решение по использованию nullFlavor.

**Reply by eugenekogan-spb** [reply]
Обсуждение пока не состоялось и решение не принималось. 
Я со своей стороны поддерживаю "вариант 1": nullFlavor в расширении адреса на уровне RuCore не нужны, по крайней мере на данном этапе его разработки, поскольку нет такой задачи.
В принципе, в мини-модель стоило бы вставить обработку nullFlavor, как пример, который потом может встретиться при профилировании. В таком случае я бы предложил рассмотреть его указание для отчества пациента. Некоторые учетные формы, в которых надо ввести ФИО пациента, содержат отдельный флажок "отчество отсутствует".

**Reply by Vrodvi** [reply]
Поддерживаю выбор 1 варианта как самого простого, и в связи с этим наиболее подходящего для формируемого минимального набора артефактов.
Также текущие описания профилей и примеров экземпляров уже сформированы в соответствии с вариантом 1.
В связи с этим, замечание можно закрывать.

**Reply by Vrodvi** [reply]
Замечание "1." закрыто.

**Comment by Vrodvi** [comment]
3. Структура идентификации по ФИАС отличается от СЭМДов

В СЭМДах для идентификации ФИАС используется комплексное расширение с двумя атрибутами - идентификатор адресного объекта и идентификатор дома. Почему структура сделана отличной от СЭМДов? Почему наименование атрибутов сделано отлично от СЭМДов (там это AOGUID, тут core-ex-address-fias)?

Вот тут в деталях уже это было разобрано и была предложена аналогичная СЭМДам комплексная структура - https://github.com/fhir-ru/core/discussions/83.

Ожидается, что структура идентификации по ФИАС будет приведена к структуре СЭМДов или будет явно обозначено обоснование расхождения и дана схема маппирования предложенной структуры на структуру СЭМДов.

**Reply by Vrodvi** [reply]
Для сведения структуры с СЭМДами можем предложить использовать следующую заготовку, которая уже ранее предлагалась в дискуссии #83 (но она требует доработки - в плане наименования артефактов, например, были новые регламенты и по ним нужно сверить наименования; также нужно учесть решения, которые будут приняты в https://github.com/fhir-ru/core/discussions/242 по поводу выбранного идентификато(ов) для пространства идентификаторов ФИАС):
```
Extension: Fhir_ru_ext_fias 
Id: fhir-ru-ext-fias
Title: "Кодирование адреса по ФИАС"
Description: "Кодирование адреса по ФИАС."
* ^url = "urn:hl7-ru:fias"

* extension contains
    auguid 1..1 MS and
    houseguid 0..1 MS

* extension[auguid].value[x] only Identifier
* extension[auguid].valueIdentifier.system = "urn:ietf:rfc:3986"
* extension[auguid].valueIdentifier.value 1..1 MS
* extension[auguid].valueIdentifier.value ^short = "GUID"

* extension[houseguid].value[x] only Identifier
* extension[houseguid].valueIdentifier.system = "urn:ietf:rfc:3986"
* extension[houseguid].valueIdentifier.value 1..1 MS
* extension[houseguid].valueIdentifier.value ^short = "GUID"
```

**Reply by eugenekogan-spb** [reply]
Исправлено на комплексное расширение с двумя атрибутами в соответствии со структурой СЭМД, по варианту двух пространств идентификаторов ФИАС.

**Reply by Vrodvi** [reply]
Замечание "3." закрыто.

**Comment by Vrodvi** [comment]
4. В представленных фрагментах кода (FSH) сбивается форматирование 

К примеру символ звёздочки превратился в обозначение элемента списка.
Для предотвращения искажений форматирования кода просьба все кодовые фрагменты оборачивать в специальную разметку - тройной символ апострофа.
Пример:
если указать:
\```
\* фрагмент кода 
\```
, то будет корректное отображение:
```
* фрагмент кода 
```
, а не 
* фрагмент кода

**Reply by eugenekogan-spb** [reply]
Исправлено.

**Reply by Vrodvi** [reply]
Замечание "4." закрыто.

**Comment by Vrodvi** [comment]
5. В карточке не хватает примера экземпляра данных, соответствующего данному расширению.

Расширение само по себе не может иметь примера экземпляра данных, поэтому следует приложить ссылку на страницу с примером экземпляра для типа данных или для пациента (в котором указывается заполнение, в том числе, адреса с данным расширением)

**Reply by eugenekogan-spb** [reply]
Включена ссылка на страницу с примером экземпляра адреса.

**Reply by Vrodvi** [reply]
Замечание "5." закрыто.

**Comment by Vrodvi** [comment]
6. Необходима страница с описанием регламента / лучших практик по формирования данного типа артефактов - "Расширений"

Была договорённость, что результатом работы по каждому типу артефакта будет, в том числе, и специальная страничка/документ с описанием особенностей/рекомендаций по формированию данного типа артефактов (особенности, которые были идентифицированы в ходе обсуждения/разработки данного типа артефакта в нашем минимальном наборе)

Некоторые особенности, которые можно указать в документе:
- различие при формировании простого расширения против формирование комплексного расширения
- особенности использования nullFlavor для атрибутов расширения и самого расширения
- особенности формирования ограничений на несколько атрибутов расширения (например, обязательное указание одного из двух атрибутов, но не обоих одновременно (XOR, "исключающее ИЛИ"))
- особенности присвоение URL для самого расширения и для его составных частей (если расширение - комплексное)

**Reply by eugenekogan-spb** [reply]
Начало положено: https://github.com/fhir-ru/core/discussions/246
Перечисленные здесь "особенности, которые можно указать в документе" можно обсудить и добавить после стабилизации стиля методических рекомендаций.

**Reply by Vrodvi** [reply]
Вынес ссылку на страницу с описанием в карточку.
Пока что этого достаточно.
Замечание "6." закрыто.

**Comment by Vrodvi** [comment]
7. Проблема указания "не-канонического" url

Поскольку для более полного соответствия СЭМДам в данном расширении был переопределён URL расширения с канонического на `^url = "urn:hl7-ru:fias"`, то требуется проверить - не приведёт ли это к появлению ошибок и замечаний в ходе формирования сайта с руководством по реализации через стандартный Publisher (ошибки и замечания на странице "QA Report").
Если ошибки/замечания будут, то требуется более детально изучить вопрос формирования URL для FHIR-артефактов и выработать рекомендации, чтобы в дальнейшем не сталкиваться с этим типом ошибок.

**Reply by Vrodvi** [reply]
Замечание не закрыто - требуется проанализировать "QA Report".

**Reply by eugenekogan-spb** [reply]
На [встрече 21.11.23](https://fhir.ru/meetings.2023-11-21) приняли во внимание следующее: 
1) публикатор дает ошибку, возможность ее подавить неизвестна
2) если вернуться к каноническому url, то можно разместить по этому адресу информацию о том, как поступать при маппинге на СЭМД. 

Решили вернуться к каноническому url и дать комментарий, что для маппинга на СЭМД должен использоваться ранее указанный urn. Согласно этому решению, заменен ^url в основном расширении и добавлен комментарий. 

System в нижестоящих расширениях не изменялся, поскольку это решение не оговаривалось!

**Reply by Vrodvi** [reply]
К сожалению, указанный `^url` отличается от канонического:
 "http://fhir.ru/core/Extension/address-fias" - такой указан,
 "http://fhir.ru/core/StructureDefinition/core-ex-address-fias" - вот такой должен быть канонический.

Для использования канонического url достаточно не указывать его, тогда он будет сформирован автоматически по правилу
[базовый url]+[тип ресурса]+[id]
(тип ресурса в данном случае - StructureDefinition, а вот типа Extension - не существует)

**Reply by eugenekogan-spb** [reply]
Ясно. Поправил. Добавил в рекомендации.
Если (до тех пор пока) не будет принято другое решение, я бы предпочел url указывать явно, потому что fsh online подставляет example.org и дает ошибку.

**Reply by Vrodvi** [reply]
В fsh online можно настроить базовый url - нажмите Configuration, далее - в поле Canonical Url указать "http://fhir.ru/core":
![image](https://github.com/fhir-ru/core/assets/92928137/56002243-38cb-4b8d-be8c-51d4b1296735)

**Reply by eugenekogan-spb** [reply]
Спасибо, сработало. ^url убран, в рекомендации поправлено.

**Reply by Vrodvi** [reply]
Замечание "7." закрыто.
